src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1214,9 +1236,56 @@ public bool StatusCanHaveBody(int statusCode)
                    statusCode != 304;
         }
 
-        private void ThrowResponseAlreadyStartedException(string value)
+        private void RejectNonBodyTransferEncodingResponse(bool appCompleted)
+        {
+            var ex = BadHttpResponseException.GetException(ResponseRejectionReason.NonBodyResponseTransferEncodingSet, StatusCode);
+            if (!appCompleted)
+            {
+                // Back out of header creation surface exeception in user code
+                _requestProcessingStatus = RequestProcessingStatus.RequestStarted;
+                throw ex;
+            }
+            else
+            {
+                ReportApplicationError(ex);
+                // 500 Internal Server Error
+                PrepareErrorResponse(statusCode: 500);
+            }
+        }
+
+        private void PrepareErrorResponse(int statusCode)
+        {
+            StatusCode = statusCode;
+            ReasonPhrase = null;
+
+            var responseHeaders = FrameResponseHeaders;
+            responseHeaders.Reset();
+            var dateHeaderValues = DateHeaderValueManager.GetDateHeaderValues();
+
+            responseHeaders.SetRawDate(dateHeaderValues.String, dateHeaderValues.Bytes);
+            responseHeaders.SetRawContentLength("0", _bytesContentLengthZero);
+
+            if (ServerOptions.AddServerHeader)
+            {
+                responseHeaders.SetRawServer(Constants.ServerName, _bytesServer);
+            }
+        }
+
+        public void HandleNonBodyResponseWrite(int count)
         {
-            throw new InvalidOperationException(value + " cannot be set, response had already started.");
+            if (Method == "HEAD")
+            {
+                // Don't write to body for HEAD requests.
+                if (Log.IsEnabled(LogLevel.Debug))
+                {
+                    Log.ConnectionHeadResponseBodyWrite(ConnectionId, count);
```

Comment:
So we're not throwing here because a lot of real apps will write responses without looking at the request method?